### ISSUE TITLE
Vue.js support?

### ISSUE THREAD

**Issue by Niedzwiedzw** [kak-lsp/kak-lsp#442]
there is a nice language server for vue, for some reason html-languageserver gets spawned whenever I open up .vue files. Any idea why would that be?

**Comment by krobelus** [comment]
If html-languageserver is started, then your `%opt{filetype}` is likely HTML.
You can probably do something like
`hook global BufCreate .*[.]vue %{ set-option buffer filetype vue }`
and then add an entry for "language.vue" in your `kak-lsp.toml`.

For syntax highlighting I don't think there's an official vue config yet, but it should be easy to write - you can reuse most of the javascript filetype.

**Comment by Niedzwiedzw** [comment]
https://github.com/kakoune-editor/kakoune-extra-filetypes/pull/6#discussion_r563872908 I'll cross post this here for future users

**Comment by Niedzwiedzw** [comment]
the highlights work more or less thanks to @krobelus  hack, the only thing  that's missing is the lsp support itself :) 
https://github.com/vuejs/vetur/issues/2648 here's the link to the issue on vue's language server repo, I've posted some debug logs there. maybe someone smarter than me can spot some low hanging fruit there to achieve at least partial functionality.

**Comment by Niedzwiedzw** [comment]
vue language server's maintainer analyzed the logs and his theory is as follows:
```
LSP client must to send params.workspaceFolders in initialize request.
If you don't send any project path, vetur will print No workspace path found. Vetur initialization failed..
```
https://github.com/vuejs/vetur/issues/2648#issuecomment-767393393

is this something kak-lsp supports?

**Comment by Niedzwiedzw** [comment]
https://github.com/kak-lsp/kak-lsp/blob/354b46e3cf56f0da35b444941a701ca4c1135aa8/src/general.rs#L78 I guess that's the place, could we expose that as an option in the toml config easily?

**Comment by Niedzwiedzw** [comment]
hmm I guess the easiest and also somewhat unixy approach would be to try to read environment variable lak KAK_LSP_WORKSPACE_FOLDER and try to put it in there if it's not absent...

**Comment by Niedzwiedzw** [comment]
https://github.com/kak-lsp/kak-lsp/pull/443/ this is a draft, I haven't tested it, but is that something that we'd be interested in?

**Comment by krobelus** [comment]
Nice, you've already found the missing piece to the puzzle :)

In the end the implementation should probably detect this automatically (like other editor's language clients).
Isn't the workspace usually the same as the project root dir?

**Comment by Niedzwiedzw** [comment]
yeah that was my thinking, since for example package.json is specified as a root file of a javascript project it could jump couple directories up to see if that file is found... not sure how to access the config at that point but it should be pretty trivial to do... But still maybe it's worth leaving that env var as an override mechanism in case some edge case isn't caught?

**Comment by krobelus** [comment]
Tangentially related: kak-lsp already reads env vars like `KAK_LSP_PROJECT_ROOT_HTML`, see `src/project_root.rs 45`

I looked at what Emacs [`lsp-mode`](https://github.com/emacs-lsp/lsp-mode) does:
- `lsp--try-project-root-workspaces` adds the project to `lsp-session-folders` (which is the variable they use for `workspaceFolders`
- `lsp-workspace-folders-add` allows the user to add new projects at runtime

Probably an environment variable is the wrong interface: I'd rather expose something like a `lsp_workspaceFolders` Kakoune config option that can be set dynamically.

> not sure how to access the config at that point

`root_path` is a parameter to `initialize` in `src/general.rs:35`. The first implementation can simpy always send that as single root path. That will probably work for most use cases. Followups can make this configurable. The config is available as part of the context parameter.

Let me know if you need help with testing, or if you have instructions how to reproduce/test this.

**Comment by Niedzwiedzw** [comment]
sorry for not responding for some time, had some work to do :D even with the hacky envvar solution of mine there are still errors

when running `lsp-list-capabilities` I'm getting 
```
Jan 29 19:49:38.358 DEBG From editor:
session  = "1121750"
client   = "client0"
buffile  = "/home/niedzwiedz/Pieprzyk/zajedz-front/src/components/menu/SauceChoice.vue"
filetype = "html"
version  = 112
method   = "capabilities"
[params]
, module: kak_lsp::editor_transport:122
Jan 29 19:49:38.358 DEBG Searching for vars starting with KAK_LSP_PROJECT_ROOT_HTML, module: kak_lsp::project_root:46
Jan 29 19:49:38.358 DEBG Routing editor request to Route { session: "1121750", language: "html", root: "/home/niedzwiedz/Pieprzyk/zajedz-front" }, module: kak_lsp::session:95
Jan 29 19:49:38.358 DEBG To editor `1121750`: eval -client client0 -verbatim -- info 'kak-lsp commands supported by html language server:

lsp-hover
lsp-completion (hooked on InsertIdle)
lsp-definition (mapped to `gd` by default)
lsp-references (mapped to `gr` by default)
lsp-code-actions
lsp-diagnostics', module: kak_lsp::editor_transport:84
```
so not much
when trying to do lsp-code-actions at this point, I'm getting

```
Jan 29 19:50:41.079 DEBG To server: {"jsonrpc":"2.0","method":"textDocument/codeAction","params":{"context":{"diagnostics":[]},"range":{"end":{"character":6,"line":31},"start":{"character":6,"line":31}},"textDocument":{"uri":"file:///home/niedzwiedz/Pieprzyk/zajedz-front/src/components/menu/SauceChoice.vue"}},"id":11}, module: kak_lsp::language_server_transport:183
Jan 29 19:50:41.079 DEBG From server: {"jsonrpc":"2.0","id":11,"error":{"code":-32601,"message":"Unhandled method textDocument/codeAction"}}, module: kak_lsp::language_server_transport:157
Jan 29 19:50:41.079 ERRO Error response from server: Failure { jsonrpc: Some(V2), error: Error { code: MethodNotFound, message: "Unhandled method textDocument/codeAction", data: None }, id: Num(11) }, module: kak_lsp::controller:155
Jan 29 19:50:41.079 DEBG To editor `1121750`: eval -client client0 -verbatim -- lsp-show-error 'html language server doesn''t support method textDocument/codeAction', module: kak_lsp::editor_transport:84
```

so seems like the missing workspace part was just the beginning...

**Comment by krobelus** [comment]
Sounds like the server doesn't support code actions. Do other things (like lsp-definition) work? Do code actions work in Emacs?

**Comment by Niedzwiedzw** [comment]
same... maybe it's about `textDocument/` prefix?
```
 Jan 30 17:02:52.555 ERRO Error response from server: Failure { jsonrpc: Some(V2), error: Error { code: MethodNotFound, message: "Unhandled method textDocument/definition", data: None }, id: Num(5) }, module: kak_lsp::controller:155
 ```
